### PR TITLE
`SpliceFlatStreamToMetaSingle`: propagate cancel when races with data

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

Cancellation of the single may race with delivering the first element. If `cancel` wins, we already deliver `CancellationException` if someone subscribes to the payload publisher. However, if `cancel` arrives after  `dataSubscriber.onSuccess(...)` and before someone subscribes to the payload publisher, it's a strong indicator that there was a race and `dataSubscriber` is not interested in it anymore. Therefore, we should pessimistically assume that nobody will ever subscribe to the payload publisher, cancel upstream subscription, and deliver `CancellationException` is someone still subscribes.

Modifications:
- Cancel upstream subscription for both `null` and `PENDING` states if received data cancel;
- Upwrap code around `parent.packer.apply` to assert that this path happens only if `CANCELED`;
- Deliver `CancellationException` is someone still subscribes to the payload publisher after `CANCELED`;
- Fix another bug: do not deliver `dataSubscriber.onError` if `metaSeenInOnNext == true` because it's already terminated (`CANCELED` does not play any role here);
- Log more appropriate message if upstream delivers a terminal event after `payloadSubscriber` is already terminated (`EMPTY_COMPLETED_DELIVERED`);
- Remove duplication around "Duplicate Subscribers are not allowed";

Result:

Network resources receive `cancel` signal and clean up connection state.

---
Note that the original version of this operator had this behavior, but it was changed in https://github.com/apple/servicetalk/commit/38e2c67ff19f67e3e9730d56e379e726e1981aa6
We don't have a full PR for that commit with review comments. Based on the commit description I believe it was fixing a different problem and the change in cancel propagation behavior for `PENDING` state is likely a side-effect of introducing a new state.

### CODE REVIEW

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/JavaNetSoTimeoutHttpConnectionFilterTest.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/JavaNetSoTimeoutHttpConnectionFilterTest.java:71]: This PR is blocked by awaiting a fix for `Single.ambWith(...)` operator

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java:161]: Please carefully review the updated state machine and existing tests that we have. If you see a use-case not covered by the tests, will appreciate new ideas.

**Review by idelpivnitskiy — COMMENTED** [review]
Ready for review, but will be merged only after `Single.ambWith(...)` fix

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java
Comment (bryce-anderson) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java:207]: I believe this `cas` should never fail: can we throw an AssertionError or similar if it does?

**Review by bryce-anderson — APPROVED** [review]
Looks good but it's complex code so I'd recommend getting one more review before merging.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java:207]: good idea, it should never fail

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java
Comment (Scottmitch) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java:337]: is this expected to be common enough where it needs to be stackless? if not, its called from 2 different locations so we should differentiate these two cases.

**Review by Scottmitch — APPROVED** [review]
lgtm after ignored test is brought back to working

**Pull request merged**
